encoding/protobuf/outbound_test.go

```diff
@@ -115,3 +122,75 @@ func TestNoResponseHeaders(t *testing.T) {
 	assert.Contains(t, err.Error(), "code:invalid-argument")
 	assert.Contains(t, err.Error(), "response headers are not supported for streams")
 }
+
+var _ jsonpb.AnyResolver = (*testResolver)(nil)
+
+func TestOutboundAnyResolver(t *testing.T) {
+	const testValue = "foo-bar-baz"
+	newReq := func() proto.Message { return &testpb.TestMessage{} }
+	customResolver := &testResolver{NewMessage: newReq}
+
+	tests := []struct {
+		name     string
+		anyURL   string
+		resolver jsonpb.AnyResolver
+		wantErr  bool
+	}{
+		{
+			name:   "nothing custom",
+			anyURL: "uber.yarpc.encoding.protobuf.TestMessage",
+		},
+		{
+			name:     "custom resolver",
+			anyURL:   "uber.yarpc.encoding.protobuf.TestMessage",
+			resolver: customResolver,
+		},
+		{
+			name:     "custom resolver, custom URL",
+			anyURL:   "foo.bar.baz",
+			resolver: customResolver,
+		},
+		{
+			name:    "custom URL, no resolver",
+			anyURL:  "foo.bar.baz",
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			trans := yarpctest.NewFakeTransport()
+			// outbound that echos the body back
+			out := trans.NewOutbound(nil, yarpctest.OutboundCallOverride(
+				yarpctest.OutboundCallable(func(ctx context.Context, req *transport.Request) (*transport.Response, error) {
+					return &transport.Response{Body: ioutil.NopCloser(req.Body)}, nil
+				}),
+			))
+
+			client := NewClient(ClientParams{
+				ClientConfig: &transport.OutboundConfig{
+					Outbounds: transport.Outbounds{
+						Unary: out,
+					},
+				},
+				AnyResolver: tt.resolver,
+				Options:     []ClientOption{UseJSON},
+			})
+
+			testMessage := &testpb.TestMessage{Value: testValue}
+
+			// convert to an Any so that the marshaller will use the custom resolver
+			any, err := ptypes.MarshalAny(testMessage)
```

Comment:
This is probably working fine because the byte representation is the same, but let's be safe and use gogo everywhere.